File: CPP/quicksort/quicksort_250000.cpp
#include <iostream>
#include <vector>
using namespace std;

void quickSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int pivot = arr[(left + right) / 2];
        int i = left, j = right;
        while (i <= j) {
            while (arr[i] < pivot)
                i++;
            while (arr[j] > pivot)
                j--;
            if (i <= j) {
                swap(arr[i], arr[j]);
                i++;
                j--;
            }
        }
        quickSort(arr, left, j);
        quickSort(arr, i, right);
    }
}

int main() {
    bool debug = false;

    int n = arr.size();
    quickSort(arr, 0, n-1);
    if(debug) {
        int n = arr.size();
        for (int i = 0; i < n; i++){
            cout << arr[i] << " ";
        }
    } 

    return 0;
}

File: Rust/quicksort/quicksort_250000.rs
fn quicksort<T: Ord + Copy>(arr: &mut [T]) {
    let len = arr.len();
    if len <= 1 {
        return;
    }
    let pivot = arr[len / 2];
    let mut i = 0;
    let mut j = len - 1;
    while i <= j {
        while arr[i] < pivot {
            i += 1;
        }
        while arr[j] > pivot {
            j -= 1;
        }
        if i <= j {
            arr.swap(i, j);
            i += 1;
            j -= 1;
        }
    }
    if j > 0 {
        quicksort(&mut arr[0..=j]);
    }
    if i < len - 1 {
        quicksort(&mut arr[i..]);
    }
}
    

fn main() {
    let debug: bool = false;
    quicksort(&mut arr);
    if debug {
        println!("{:?}", arr);
    }
}